plugins/module_utils/foreman_helper.py

```diff
@@ -285,14 +295,132 @@ def ensure_resource(self, resource, entity_dict, entity, state=None, name_map=No
+                changed, current_entity = self._create_entity(resource, desired_entity, params, entity_spec)
+            else:
+                changed, current_entity = self._update_entity(resource, desired_entity, current_entity, params, entity_spec)
+        elif state == 'absent':
+            if current_entity is not None:
+                changed, current_entity = self._delete_entity(resource, current_entity, params)
+        else:
+            self.fail_json(msg='Not a valid state: {}'.format(state))
+        return changed, current_entity
+
+    def _create_entity(self, resource, desired_entity, params, entity_spec):
+        """Create entity with given properties
+
+            Parameters:
+                resource (string): Plural name of the api resource to manipulate
+                desired_entity (dict): Desired properties of the entity
+                params (dict): Lookup parameters (i.e. parent_id for nested entities) (optional)
+                entity_spec (dict): Description of the entity structure
+            Return value:
+                Pair of boolean indicating whether something changed and the new current state if the entity
+        """
+        payload = _flatten_entity(desired_entity, entity_spec)
+        if not self.check_mode:
+            if params:
+                payload.update(params)
+            return self._resource_action(resource, 'create', payload)
+        else:
+            fake_entity = desired_entity.copy()
+            fake_entity['id'] = -1
+            return True, fake_entity
+
+    def _update_entity(self, resource, desired_entity, current_entity, params, entity_spec):
+        """Update a given entity with given properties if any diverge
+
+            Parameters:
+                resource (string): Plural name of the api resource to manipulate
+                desired_entity (dict): Desired properties of the entity
+                current_entity (dict): Current properties of the entity
+                params (dict): Lookup parameters (i.e. parent_id for nested entities) (optional)
+                entity_spec (dict): Description of the entity structure
+            Return value:
+                Pair of boolean indicating whether something changed and the new current state if the entity
+        """
+        payload = {}
+        desired_entity = _flatten_entity(desired_entity, entity_spec)
+        current_entity = _flatten_entity(current_entity, entity_spec)
+        for key, value in desired_entity.items():
+            # query entity_spec[key] for special treatment
+            spec = entity_spec.get(key)
+            if spec:
+                # properties with 'force_update' must always be updated
+                if spec.get('force_update', False):
+                    payload[key] = value
+                    continue
+                # properties with 'check_type' ???
+                # if spec.get('check_type', False):
+                #     # TODO
+                #     pass
```

Comment:
I do not fully understand, how `check_type` is supposed to work, but is only used in the settings module. And we might be able to so something even more clever there, because the entity provided by the server contains a hint to the type of the setting.